# 모델 학습
현재까지 진행사항
1. Training 폴더의 img 파일을 읽어와서 json 파일을 기준으로 전처리
    - 나이/성별 csv 파일 생성(labels.csv)
    - 얼굴 crop 후 저장(로컬 cropped_faces 폴더)
2. cropped_faces 폴더 불러와서
    - 성별 + 나이 클래스 조합 -> 클래스명으로 생성(class_name) -> 클래스명 -> 고유 인덱스 매핑
    - stratify를 위해 class_name 컬럼 기준으로 분리(YOLO_cls_dataset -> train/val 0.2로)
    - 해당 경로로 이미지 복사해옴
3. 로컬에서 작업완료  
    - 코랩으로 불러온 파일들: YOLO_cls_dataset.zip(labels.csv, YOLO_cls_dataset)

In [ ]:
!unzip "/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset.zip" -d "/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project"

스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0688_IND_M_45_-90_02.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0688_IND_M_45_45_02.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0690_IND_M_41_-45_02.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0690_IND_M_41_-90_01.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0690_IND_M_41_0_01.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0690_IND_M_41_45_02.jpg  
  inflating: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val/3/F0693_IND_M_40_-45_01.jpg  
  

In [ ]:
!pip install ultralytics

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 62.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 65.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 36.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 51.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 104.1 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling

## 모델 학습

In [ ]:
from ultralytics import YOLO

model = YOLO("yolov8s-cls.pt")
model.train(
    data="/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset",
    imgsz=224,
    epochs=10,
    batch=32,
    patience=3,
    name="age_sex_cls_colab"
)

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


100%|██████████| 12.3M/12.3M [00:00<00:00, 48.6MB/s]


Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=32, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=224, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8s-cls.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=age_sex_cls_colab, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overl

100%|██████████| 5.35M/5.35M [00:00<00:00, 173MB/s]


AMP: checks passed ✅
train: Fast image access ✅ (ping: 1.5±1.6 ms, read: 2.6±1.5 MB/s, size: 7.0 KB)


train: Scanning /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/train... 34542 images, 0 corrupt: 100%|██████████| 34542/34542 [03:49<00:00, 150.49it/s]


train: New cache created: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/train.cache
val: Fast image access ✅ (ping: 0.9±0.4 ms, read: 0.4±0.2 MB/s, size: 7.1 KB)


val: Scanning /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val... 10234 images, 0 corrupt: 100%|██████████| 10234/10234 [01:04<00:00, 159.74it/s]


val: New cache created: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val.cache
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000625, momentum=0.9) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.0005), 27 bias(decay=0.0)
Image sizes 224 train, 224 val
Using 2 dataloader workers
Logging results to runs/classify/age_sex_cls_colab
Starting training for 10 epochs...

      Epoch    GPU_mem       loss  Instances       Size


       1/10     0.695G      2.544         32        224:   0%|          | 4/1080 [00:02<08:12,  2.19it/s]

       1/10     0.734G      2.503         32        224:   1%|          | 7/1080 [00:05<11:40,  1.53it/s]
100%|██████████| 755k/755k [00:00<00:00, 16.9MB/s]
               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:04<00:00,  2.49it/s]

                   all      0.766      0.996



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:05<00:00,  2.43it/s]

                   all      0.809      0.999



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:05<00:00,  2.45it/s]

                   all      0.853      0.999



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:04<00:00,  2.48it/s]

                   all      0.881          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:04<00:00,  2.47it/s]

                   all      0.896          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:01<00:00,  2.59it/s]

                   all      0.918          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:02<00:00,  2.55it/s]

                   all      0.933          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:01<00:00,  2.58it/s]

                   all      0.943          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:02<00:00,  2.54it/s]

                   all       0.95          1



      Epoch    GPU_mem       loss  Instances       Size


      10/10      1.03G      0.205         14        224: 100%|██████████| 1080/1080 [05:36<00:00,  3.21it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:02<00:00,  2.58it/s]


                   all      0.953          1

10 epochs completed in 1.132 hours.
Optimizer stripped from runs/classify/age_sex_cls_colab/weights/last.pt, 10.3MB
Optimizer stripped from runs/classify/age_sex_cls_colab/weights/best.pt, 10.3MB

Validating runs/classify/age_sex_cls_colab/weights/best.pt...
Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
YOLOv8s-cls summary (fused): 30 layers, 5,090,572 parameters, 0 gradients, 12.5 GFLOPs
train: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/train... found 34542 images in 12 classes ✅ 
val: /content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/YOLO_cls_dataset/val... found 10234 images in 12 classes ✅ 
test: None...


               classes   top1_acc   top5_acc: 100%|██████████| 160/160 [01:08<00:00,  2.35it/s]


                   all      0.953          1
Speed: 0.1ms preprocess, 0.4ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to runs/classify/age_sex_cls_colab


ultralytics.utils.metrics.ClassifyMetrics object with attributes:

confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e15cb90fc50>
curves: []
curves_results: []
fitness: 0.97645103931427
keys: ['metrics/accuracy_top1', 'metrics/accuracy_top5']
results_dict: {'metrics/accuracy_top1': 0.95290207862854, 'metrics/accuracy_top5': 1.0, 'fitness': 0.97645103931427}
save_dir: PosixPath('runs/classify/age_sex_cls_colab')
speed: {'preprocess': 0.07491402003140449, 'inference': 0.3925210716248088, 'loss': 0.0004831612273226318, 'postprocess': 0.0005792697857556513}
task: 'classify'
top1: 0.95290207862854
top5: 1.0

> ### 📌 학습 결과
Top-1 Acc: 95.29%  
Top-5 Acc: 100%  
Best Model Path: `runs/classify/age_sex_cls_colab/weights/best.pt`  
Loss 감소: Epoch 1 = 2.5 -> Epoch 10: 0.205  
Early Stopping: 적용X

## 모델 저장

In [ ]:
!cp -r /content/runs/classify/age_sex_cls_colab \
      "/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/saved_model"

모델 저장완료! 다음에 불러올때는,  
```python
from ultralytics import YOLO

# 저장해둔 best.pt 불러오기
model = YOLO("/content/drive/MyDrive/Colab Notebooks/vision_project/age_sex_cls_project/saved_model/age_sex_cls_colab/weights/best.pt")

# 예측 예시 (단일 이미지 혹은 폴더 가능)
results = model.predict(source="/content/test_image.jpg", imgsz=224)

# 전체에 대해 추론하려면
#results = model.predict(source="/content/test_images_folder", imgsz=224, save=True)  
```